### ISSUE TITLE
Segfault installing via pip3 on Raspberry Pi 3B

### ISSUE THREAD

**Issue by doug-e7systems** [paramiko/paramiko#1185]
Hi:

The following produces a segfault 100% of the time I try it.  I tried this on 2 different Raspberry Pi 3B computers at least 6 times.

This is simplified from a larger install.
To reproduce, get the latest Raspbian image (2018-03-13-raspbian-stretch.zip) and boot the Pi with it.

Then ssh into the box and run
```
pi@raspberrypi:~ $ pip3 install cryptography
pi@raspberrypi:~ $ pip3 install paramiko
Segmentation fault
```

The log below is from running
`pip3 -vvv install paramiko`

[paramiko_log.txt](https://github.com/paramiko/paramiko/files/1858582/paramiko_log.txt)

I am not sure if this is paramiko, cryptography or the Raspberry Pi.  Any pointers would be greatly appreciated. 

Thanks!

**Comment by doug-e7systems** [comment]
A quick progress update.  I got the Pi to dump a core file, I am trying to get the python debug symbols to load so I can figure out more...

This is a weird bug, it does not occur on x86, only on ARM.  Pip actually generates a valid whl, which I can use for installs just fine.  But if I let pip try and generate the whl, that's when I get the segfault.

Really weird.  Anyhow, I'll let you know how I get on.

**Comment by Russell-Jones** [comment]
I tried the following on an RPi 3 (I used a virtual env as I don't want random stuff in ~/.local or /usr/local that I'll forget about and get confused by later)
```
python3.5 -m venv --system-site-packages 35ve
cd 35ve/
source bin/activate
python3.5 -m pip install --upgrade pip setuptools wheel
which pip # should output /path/to/35ve/bin/pip
pip install --upgrade cryptography
pip install --upgrade paramiko
```
I got the same weird output:
```
Successfully built pynacl bcrypt
Installing collected packages: pycparser, cffi, six, pynacl, enum34, asn1crypto, idna, ipaddress, cryptography, bcrypt, pyasn1, paramiko
Successfully installed asn1crypto-0.24.0 bcrypt-3.1.4 cffi-1.11.5 cryptography-2.2.2 enum34-1.1.6 idna-2.6 ipaddress-1.0.19 paramiko-2.4.1 pyasn1-0.4.2 pycparser-2.18 pynacl-1.2.1 six-1.11.0
Segmentation fault
```
It's not clear to me what is segfaulting, but I think it may be the python3 interpreter running pip.
OK, bcrypt and pynacl get built, so maybe those?

```
pip uninstall bcrypt ; pip install bcrypt
...
Collecting bcrypt
Collecting cffi>=1.1 (from bcrypt)
Collecting six>=1.4.1 (from bcrypt)
  Using cached six-1.11.0-py2.py3-none-any.whl
Collecting pycparser (from cffi>=1.1->bcrypt)
  Using cached https://www.piwheels.org/simple/pycparser/pycparser-2.18-py2.py3-none-any.whl
Installing collected packages: pycparser, cffi, six, bcrypt
Successfully installed bcrypt-3.1.4 cffi-1.11.5 pycparser-2.18 six-1.11.0
Segmentation fault
```

Huh. OK, does pynacl do this? Yes
```
Installing collected packages: six, pycparser, cffi, pynacl
Successfully installed cffi-1.11.5 pycparser-2.18 pynacl-1.2.1 six-1.11.0
Segmentation fault
```
But wait! This stuff is all in .local/lib/python2.7! `rm -r ~/.local/lib/python2.7/`

OK, try again
```
python3.5 -m pip install --upgrade cryptography
python3.5 -m pip install --upgrade paramiko
```
Wooh! No segfault or guff in ~/.local/lib

Postscript:

`/path/to/35ve/bin/pip install --upgrade cryptography` does the right thing
`pip install --upgrade cryptography` does not, despite what `which pip` says
```
$ pip --version
pip 9.0.1 from /usr/lib/python2.7/dist-packages (python 2.7)
$ pip3 --version
pip 9.0.3 from /path/to/35ve/lib/python3.5/site-packages (python 3.5)
$ hash -r
$ pip --version
pip 9.0.3 from /home/pi/Documents/python/35ve/lib/python3.5/site-packages (python 3.5)
```
Yeuch. Well, live and learn. `which`, from debianutils 4.8.1.1 at least, seems to lie to you under certain circumstances (I suspect if a command is used before a PATH change)

**Comment by ploxiln** [comment]
`which` looks up where an executable would be found given the current PATH and currently existing executables. but the shell caches where executables were found in the past, until you run `hash -r` (a shell built-in) to clear that cache.

When you `source bin/activate` it (probably) clears the cache for you as part of that shell include. But when you run `python3.5 -m pip install --upgrade pip` the shell's cache is not cleared. By the way, inside the virtual-env, you should just use "python" not "python3.5"

**Comment by Russell-Jones** [comment]
Right, that makes sense as a cause. As a matter of habit, I prefer to specify the Python version in the interpreter name; at least saying python3 or python2, as per PEP 394. Why would that be contraindicated in a virtualenv?

@doug-e7systems  Where does the segfault show up? Is it right at the end as I saw?

**Comment by ploxiln** [comment]
I figured that always using just "python" inside the virtualenv was more fool-proof, it will always get you the wrapper inside the virtualenv ... whereas if you accidentally specify a version-alias python which is not setup inside the virtualenv you can get a python outside the virtualenv without warning. I actually wasn't sure all the version aliases for the selected python were set-up in the virtualenv, but I see they are. I guess it's not a big deal :)

**Comment by cheeseandcereal** [comment]
I was actually having a similar problem (although not on a raspi, it was on an x86_64 intel machine running ubuntu 18.04).

I narrowed it down to the cffi package (see attached log). It only crashes after everything has been installed and pip is cleaning up. Not sure why or what I could do to further debug, as I'm not sure what's responsible for the pip cleanup.
[cffi_install_pip_log.txt](https://github.com/paramiko/paramiko/files/2207584/cffi_install_pip_log.txt)
